engine/source_sym_engine.ml

```diff
@@ -0,0 +1,184 @@
+    | Leaf (pilist, expr) ->
+      bprintf buf
+        "Leaf='%a':\
+         %t"
+        bprint_source_expr expr
+        (break (ntabs+1)) ;
+        bprintf buf "constraints {%a}" (bprint_list ~sep bprint_pi) pilist
+    | Node (pi, cst_list) ->
+      bprintf buf
+         "Node: constraints {%a}\
+          %t%a"
+         bprint_pi pi
+         (break (ntabs+1))
+         (bprint_list ~sep (bprint_tree (ntabs+1))) cst_list
+    | Conjunction (cst_list) ->
+      bprintf buf
+         "Conjunction:%t%a"
+         (break (ntabs+1))
+         (bprint_list ~sep (bprint_tree (ntabs+1))) cst_list
+    | Root cst_list ->
+       bprintf buf
+         "Root=\
+          %t%a"
+         (break (ntabs+1))
+         (bprint_list ~sep (bprint_tree (ntabs+1))) cst_list
+  in
+  let buf = Buffer.create 42 in
+  bprint_tree 0 buf stree;
+  BatIO.write_line BatIO.stdout (Buffer.contents buf)
+
+let sym_exec (source: source_program) =
+  let shares_kst ?wildcard_to_none:(w=false) k (clause:clause): clause option =
+    let pattern, expr = clause in 
+    match pattern with
+    | Wildcard when w = false-> Some (Wildcard, expr)
+    | Constructor (k', plist) when k' = k ->
+      if (List.length plist > 1) then
+        Some (Constructor (Tuple, plist), expr)
+      else
+        Some (plist |> List.hd, expr)
+    | _ -> None
+  in
+  let negate_piop = function
+    | EqInt i -> NeqInt i
+    | NeqInt i -> EqInt i
+    | IsK s -> NotK s
+    | NotK s -> IsK s
+    | IsTrue -> IsFalse
+    | IsFalse -> IsTrue
+    | IsTuple -> NotTuple
+    | NotTuple -> IsTuple
+    | IsNil -> NotNil
+    | NotNil -> IsNil
+    | IsCons -> NotCons
+    | NotCons -> IsCons
+    | EqString s -> NeqString s
+    | NeqString s -> EqString s
+  in
+  let piop_of_constructor = function
+    | Variant s -> IsK s
+    | Int i -> EqInt i
+    | Bool b -> if b then IsTrue else IsFalse
+    | String s -> EqString s
+    | Tuple -> IsTuple
+    | Nil -> IsNil
+    | Cons -> IsCons
+  in
+  let rec eval_clauses ?idx:(idx=0) constraints clauses : constraint_tree list =
+    match clauses with
+    | clause::tl -> let pattern, exp = clause
+      in
+      begin match pattern with
+        | Or (p1, p2) ->
+          eval_clauses constraints ((p1,exp)::(p2,exp)::tl) 
+        | Constructor (k, plist) ->
+          let pi = {idx=idx; op=piop_of_constructor k} in
+          let notpi = {pi with op=negate_piop pi.op} in 
+          let constraints' = (notpi :: constraints) in
+          begin match k with
+            | Int _ | String _ | Bool _ ->
+              assert (plist = []);
+              Leaf ([pi], exp)::(eval_clauses constraints' tl)
+            | Cons | Tuple -> assert (plist <> []);
+              let flat_children = plist |>
+                                  List.mapi (fun i pattern -> eval_clauses ~idx:i [] [(pattern, exp)]) |>
+                                  BatList.n_cartesian_product |>
+                                  List.map (fun n -> Conjunction n)
```

Comment:
The way standard pattern-decomposition algorithms work, you don't need a cartesian product. Instead what we do is that we use the matrix form: we keep decomposing on the first row of the matrix (and submatrices, etc.), so this corresponds to first analyzing the first argument (and its subarguments), then eventually the second argument, then the third argument, etc. The traversal will naturally perform the sort of case duplication that happens in your cartesian product here.

Here you analyze each argument independently, with the same `exp` right-hand-side (the one of the whole expression), and then you do a cartesian-product to compute a conjunction of all of that. It might be correct (I don't understand yet what happens when there are several right-hand-sides possible, for example if there are two lines in the first column of the matrix that use a `Cons` or `Tuple` pattern; you don't seem to do any grouping of lines in this case), but it requires to introduce the `Conjunction` constructor which is fairly weird, in particular because it's not clear what it means for right-hand-sides (the `exp` part) if they different in different sides of the conjunctions.

(I would be in favor of using the standard approach, but we can discuss this point more to find out whether yours actually work.)